corehq/messaging/tasks.py

```diff
@@ -146,4 +150,17 @@ def run_messaging_rule(domain, rule_id):
     # complete at about the time that the last tasks are finishing up.
     # This beats saving the task results in the database and using a
     # celery chord which would be more taxing on system resources.
-    set_rule_complete.delay(rule_id)
+    queue_task_with_retries(set_rule_complete, rule_id)
+
+
+def queue_task_with_retries(task_to_queue, *args, **kwargs):
+    for attempt in range(5):
+        try:
+            task_to_queue.delay(*args, **kwargs)
+            # break from loop if no issues
+            break
+        except SoftTimeLimitExceeded:
```

Comment:
Are you sure this will work?  Is the failure raised while queuing the task or inside the task itself?  I'd expect this to queue the task to run in another process, and then _that_ process would see the actual exception.

Here's an example that does basically what I think you want here (including waiting between retries)
https://github.com/dimagi/commcare-hq/blob/acf534eaa3561ae0038cf6d5eb531e77183a70b2/corehq/apps/users/tasks.py#L215-L232

Reply:
I think that that retry would re-run the method if it fails but here the error happened even before the task `sync_case_for_messaging_rule` was registered to run

Reply:
Makes sense after seeing https://sentry.io/organizations/dimagi/issues/1631921255/events/464d3217fe9546f48723bc509411ac38/?project=1545828&statsPeriod=14d
Very strange error - looks like it's been raised only a few times, but not only for this task, FWIW:
https://sentry.io/organizations/dimagi/issues/?project=1545828&query=is%3Aunresolved+SoftTimeLimitExceeded&statsPeriod=14d

Reply:
Actually, looking a little more at what that exception means, I'm not sure this is the correct behavior

https://docs.celeryproject.org/en/stable/userguide/workers.html#time-limits

It sounds like `run_messaging_rule` is exceeding the task time limit we've set here:

https://github.com/dimagi/commcare-hq/blob/acf534eaa3561ae0038cf6d5eb531e77183a70b2/settings.py#L548-L550

I think this needs to be somehow exempted from that limit, or split up into chunks in some way.  Since this is a "soft" limit, this approach of catching the error and retrying might actually work, though I'm not sure how best to handle it.

Reply:
thanks for pointing to that @esoergel 
Reading a bit [here](https://docs.celeryproject.org/en/stable/userguide/workers.html#time-limits) and that we don't have a hard limit set.
Also helps understand the other exceptions occurred but at different places.
And that we should not retry at all.